### ISSUE TITLE
Reject exception doesn't cause the task to fail

### ISSUE THREAD

**Issue by mhnagaoka** [celery/celery#2944]
I have a simple worker that rejects the task in case of an exception (such as `ZeroDivisionError`):

```
@app.task(bind=True, max_retries=0)
def invert(*args, **kwargs):
    try:
        x = float(kwargs['x'])
        logging.info("x=%f", x)
        result = 1.0 / float(x)
        logging.info("Result=%f", result)
    except Exception as e:
        raise Reject(e) from e
```

By running this worker and monitoring the outcome of the tasks using Celery Flower, I noticed that with Celery 3.1.16, the `Reject` exception causes the task to fail (`FAILURE` state), but with Celery 3.1.19 (actually 3.1.17+), the same exception does not change the task's state (it is kept in `STARTED` state).

I put together a simple project to help reproducing the problem: https://github.com/Arizona-Tecnologia/celery-lab

Please let me know if this is "by design" or was an unintended regression (perhaps related to #2043 ?)

**Comment by mhnagaoka** [comment]
With Celery 3.1.16, the rejected task ends in `FAILURE` state.

![image](https://cloud.githubusercontent.com/assets/903837/11515754/a85c0e34-9866-11e5-9574-03aa9879f7d0.png)

With Celery 3.1.19, the rejected task is kept in `STARTED` state.

![image](https://cloud.githubusercontent.com/assets/903837/11515792/dcd20ce0-9866-11e5-9190-42925cc63b02.png)

**Comment by ask** [comment]
FAILURE is a final state, and rejecting a task means to requeue the task, or to let it be moved to a dead letter queue (with the possibility of ignoring the message completely), so the task must still be in a pending state.

I added a new task-rejected event for 4.0, so that it can at least be visible in the monitor.

**Comment by ask** [comment]
Closed by 8454428b0f17ef4fc2c6e61614eee8a9b539c371
